src/lib/salad/hll.h

```diff
@@ -0,0 +1,158 @@
+	 * should be used for esimating large cardinalities.
+	 */
+	HLL_DENSE
+};
+
+enum {
+	/** Precision of sparsely represented HyperLogLog. */
+	HLL_SPARSE_PRECISION = 26,
+};
+
+/**
+ * Estimator that is used for the HyperLogLog algorithm.
+ * The algorithm allows to estimate cardinality of a multiset using fixed amount
+ * of memory or even less. Memory requirements and estimation accuracy are
+ * determined by the algorithm precision parameter.
+ * For the dense representation the relative error is 1.04/sqrt(m) and the
+ * memory capasity is m*6 bits where m is number of counters wich
+ * equals to 2^precision.
+ * For the sparse representation the memory usage is proportional to the number
+ * of distinct elements that has added until it reaches the memory usage of the
+ * dense representation, and then it switches to the dense representation with
+ * fixed memory requretments. The sparse representation has the best accuracy.
+ */
+struct hll {
+	/** See the comment to HLL_REPRESENTATION enum. */
+	enum HLL_REPRESENTATION representation;
+	/**
+	 * Interpretation of the data depends on the representation.
+	 * For dense representation it's an array of registers of size
+	 * 2^precision * 6 bits. Registers store the maximum added rank of set
+	 * of hashes wich last precision bits are equal to register index.
+	 * For the sparse representation it's a sparsely represented
+	 * HyperLogLog. Instead of registers the sparse representation works
+	 * with pairs consisting of the rank and the index of the added hashes.
+	 */
+	uint8_t *data;
+	/**
+	 * Precision is equal to the number of bits that are interpreted as
+	 * a register index. Available values are from HLL_MIN_PRECISION to
+	 * HLL_MAX_PRECISION (defined in hll_emprirical.h.)
+	 * The larger value leads to less estimation error but larger
+	 * memory requirement (2^precision * 6 bits).
+	 * This value is only used by the dense representation and for switching
+	 * from the sparse representation.
+	 */
+	uint8_t dprecision;
+	/**
+	 * Cached value of the last estimation.
+	 */
+	double cached_estimation;
+};
+
+/**
+ * Create a HyperLogLog estimator.
+ *
+ * \param precision defines the estimation error and memory requirements of
+ * the dense representation. The algorithm needs no more than
+ * 2^precision * 6 bits. The error formula of the dense representation is
+ * 1.04 / sqrt(2^precision). The error of the sparse representation is less than
+ * the error of the dense representation with HLL_SPARSE_PRECISION precision.
+ * The valid values are defined by constants HLL_MIN_PRECISION and
+ * HLL_MAX_PRECISION from hll_emprirical.h file.
+ *
+ * \param representation determines the data storage scheme for small
+ * cardinalities. For big carinalities the dense representation is always used.
+ * For more details see HLL_REPRESENTATION enum.
+ *
+ * \return HyperLogLog object.
+ */
+struct hll *
+hll_new(uint8_t precision, enum HLL_REPRESENTATION representation);
```

Comment:
As a potential user, I'm not sure which one I should pass here - sparse or dense - and TBH I don't want to go deep into details and do some math to figure it out. Since the implementation switches to the dense representation automatically when a certain threshold is exceeded, maybe we should remove this argument and always start sparse?

Reply:
Removed representation argument from `hll_new` function and added a new `hll_new_concrete` function with this argument. This function is used for more accurate testing and benchmarking. Also it can be used by users familiar with the algorithm.